lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -744,49 +904,179 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+
+      // Packing block is deduced by reduction dims. If all dims are reduced,
+      // then we pack in the outer most block (kernel main). Also, if all dims
+      // are reduced we pack only single tile.
+      Block *packingBlock =
+          numReducedDims == numLoopRegions
+              ? computeBlock
+              : &loopNest.loopRegions[numLoopRegions - 1 - numReducedDims]
+                     ->getBlocks()
+                     .front();
+      OpBuilder packingBuilder(packingBlock, packingBlock->begin());
+
+      packingBuilder.create<ttkernel::TileRegsAcquireOp>(
+          computeBlock->front().getLoc());
+
+      Value packSingleTile = i32(0, packingBuilder);
+      Value packingTileIndex =
+          numReducedDims == numLoopRegions
+              ? packSingleTile
+              : loopNest.loops[numLoopRegions - 1 - numReducedDims]
+                    .getRegionIterArgs()[blockArgIteratorMapping[numDPSInputs]];
+
+      packingBuilder.setInsertionPoint(packingBlock->getTerminator());
+      packingBuilder.create<ttkernel::TileRegsCommitOp>(
           computeBlock->front().getLoc());
-      innerLoopBuilder.create<ttkernel::TileRegsWaitOp>(
+      packingBuilder.create<ttkernel::TileRegsWaitOp>(
           computeBlock->front().getLoc());
-      innerLoopBuilder.create<ttkernel::PackTileOp>(
-          computeBlock->front().getLoc(), dstIndex, output,
-          iterators[blockArgIteratorMapping[numDPSInputs]]);
-      innerLoopBuilder.create<ttkernel::TileRegsReleaseOp>(
+      packingBuilder.create<ttkernel::PackTileOp>(
+          computeBlock->front().getLoc(),
+          i32(0, packingBuilder) /* dst index */, output, packingTileIndex);
+      packingBuilder.create<ttkernel::TileRegsReleaseOp>(
           computeBlock->front().getLoc());
     }
   }
 
-  SmallVector<Type>
-  getBlockArgumentTypesAsCBs(mlir::OperandRange dispatchOperands,
-                             mlir::Block::BlockArgListType blockArguments,
-                             std::int64_t numDPSInputs,
-                             PatternRewriter &rewriter) const {
+  struct StreamedOperand {
+    uint64_t srcAddress;
+    uint64_t dstAddress;
+    size_t blockArgIndex;
+    llvm::MapVector<PhysicalCoreCoord,
+                    SmallVector<TTIRToTTMetalLayoutRewriter::NocTx>>
+        dataMovement;
+    uint64_t numTiles;
+    PhysicalCoreCoordMapping coordMappping;
+
+    StreamedOperand(
+        uint64_t srcAddress, uint64_t dstAddress, size_t blockArgIndex,
+        llvm::MapVector<PhysicalCoreCoord,
+                        SmallVector<TTIRToTTMetalLayoutRewriter::NocTx>>
+            dataMovement,
+        uint64_t numTiles, const PhysicalCoreCoordMapping &coordMappping)
+        : srcAddress(srcAddress), dstAddress(dstAddress),
+          blockArgIndex(blockArgIndex), dataMovement(dataMovement),
+          numTiles(numTiles), coordMappping(coordMappping) {}
+  };
+
+  SmallVector<Type> getBlockArgumentTypesAsCBs(
```

Comment:
Should this return a pair?

Reply:
Do you mean returning `streamedOperands` as return value instead of modifying argument by ref?

Reply:
Yeah